srcbpatch/streamreplacer.cpp

```diff
@@ -0,0 +1,544 @@
+    }
+
+protected:
+    // our pairs sorted by priority - only one of them could be replaced for concrete pos
+    vector<ChoiceReplacerPair> rpairs_;
+
+    mutable size_t cachedAmount_ = 0; // we cache this amount of data
+    mutable size_t indexOfCached_ = 0; // at this index from rpairs_
+
+    // this is used to hold temporary data while the logic is 
+    // looking for the new beginning of the cached value
+    mutable vector<char> buffer_;
+};
+
+
+void ChoiceReplacer::DoReplacements(const char toProcess, const bool aEod) const
+{
+    if (nullptr == pNext_)
+    {
+        throw logic_error("Replacement chain has been broken. Communicate with maintainer");
+    }
+
+    // no more data
+    if (aEod)
+    {
+        while (cachedAmount_ > 0)
+        {
+            // to check if we send something or not
+            const size_t initialCached = cachedAmount_;
+
+            // for the rest of pairs
+            for (size_t i = indexOfCached_; i < rpairs_.size(); ++i)
+            {
+                auto [partial, full, length, index] = CheckMatch(i);
+                if (full)
+                { // need to send replacement
+                    const auto& rpair = rpairs_[index];
+                    for (size_t q = 0; q < rpair.trg_.size(); ++q) { pNext_->DoReplacements(rpair.trg_[q], false); }
+
+                    const size_t szProcessed = rpair.src_.size(); // matched source
+                    shift_left(buffer_.data(), buffer_.data() + cachedAmount_, szProcessed);
+
+                    cachedAmount_ -= szProcessed;
+                    break;
+                }
+            }
+            indexOfCached_ = 0; // now count from zero only
+
+            if (initialCached == cachedAmount_)
+            {
+                // send 1 byte
+                pNext_->DoReplacements(buffer_.data()[0], false);
+                shift_left(buffer_.data(), buffer_.data() + cachedAmount_, 1);
+                --cachedAmount_;
+            }
+        };
+
+        pNext_->DoReplacements(toProcess, aEod); // send end of the data further
+        return;
+    } // if (aEod)
+
+
+    // set buffer of cached at once
+    buffer_.data()[cachedAmount_++] = toProcess;
+    while (cachedAmount_ > 0)
+    {
+        // for the pairs
+        for (size_t i = indexOfCached_; i < rpairs_.size(); ++i)
+        {
+            auto [partial, full, length, index] = CheckMatch(i);
+            if (full)
+            { // need to send replacement
+                const auto& rpair = rpairs_[index]; // send target at once
+                for (size_t q = 0; q < rpair.trg_.size(); ++q) { pNext_->DoReplacements(rpair.trg_[q], false); }
+
+                // what if the cached index other and length is less than cached
+                shift_left(buffer_.data(), buffer_.data() + cachedAmount_, length);
+                cachedAmount_ -= length;
+                indexOfCached_ = 0; // start from the very first pair of the lexemes again
+                return;
```

Comment:
well, it could be. (I will consider)
 What are expenses of this call? Might be inline one? of define?
shift_left is already a function. The rest is 2 simple math operations

Reply:
I think it have to be inline.